### ISSUE TITLE
"When Exception" hook is needed because DBT has no exception handling

### ISSUE THREAD

**Issue by ddumasCT** [dbt-labs/dbt-core#7011]
### Is this your first time submitting a feature request?

- [X] I have read the [expectations for open source contributors](https://docs.getdbt.com/docs/contributing/oss-expectations)
- [X] I have searched the existing issues, and I could not find an existing issue for this feature
- [X] I am requesting a straightforward extension of existing dbt functionality, rather than a Big Idea better suited to a discussion

### Describe the feature

DBT has no excepton handling capabilities.  For all those who have worked with database stored procedures, this is a must, and that is why all databases have it, including cloud databases.  

The below is an excerpt from a snowflake stored procedure I wrote, which works flawlessly. It backs up data, then deletes old pre-existing data and then inserts new data. If an error occurs during delete/insert, it restores the data from backup, thus enabling idempotency. Note: The delete + insert is not a merge. It might delete 100 rows and then insert 85. Or visa versa.

If this were dbt-ized, instead of using a stored procedure, I would need to acheive the same level of data protection.  To do that, DBT needs "when exception" hooks.   In my use case, that would enable the restore code ("swap with" code).

Thanks,

Dave

– backup data in the table. snowflake has unique commands for this…  
--This could be donw in a dbt pre-hook:

drop table if exists FINANCE.CONFIGURATION_HIST_BAK;
create table FINANCE.CONFIGURATION_HIST_BAK clone FINANCE.CONFIGURATION_HIST;

– delete old and then inert new (think replace not merge)incomming data. Cannot use merge specifically
-- In DBT, I am thinking could first insert the new rows, and then use where logic in a post-hook to delete the old ones.

BEGIN

– delete -pre-existing data
DELETE FROM FINANCE.CONFIGURATION_HIST
WHERE CLOSING_MONTH = :VAR_CLOSING_MONTH;
–insert new data
INSERT INTO FINANCE.CONFIGURATION_HIST
(COUNTRY_CODE, SUBSIDIARY, COMBO, ENTRY_NO, CLOSING_MONTH)
SELECT COUNTRY_CODE, SUBSIDIARY, COMBO, ENTRY_NO, :VAR_CLOSING_MONTH
FROM FINANCE.CONFIGURATION_TEMP;

-- HERE IS THE BIG ONE BELOW.  I NEED TO EXECUTE THIS CODE ON ERROR, BUT DBT HAS NO EXCEPTON HANDLING.
-- A "WHEN EXCEPTION" HOOK COUJLD ACHEIVE THIS SAME FUNCTIONALITY:

EXCEPTION

WHEN OTHER THEN

– restore table data on any error
  ALTER TABLE FINANCE.CONFIGURATION_HIST SWAP WITH
  FINANCE.CONFIGURATION_HIST_BAK;
  VAR_BUS_ERROR := sqlerrm || ‘: Error Loading
  ENTRY_NO_CONFIGURATION_HIST’;
  raise BUS_ERROR_EXC;

END;

### Describe alternatives you've considered

Not using DBT, and using stored procedure instead.

### Who will this benefit?

Anyone (everyone) that needs exception handling capabilities.

**Comment by jtcohen6** [comment]
This deserves a more thorough response. Just jumping in for now to link a related issue/discussion from a few years ago: https://github.com/dbt-labs/dbt-core/issues/1968

**Comment by dbeatty10** [comment]
Thanks for reaching out @ddumasCT !

I've got two separate responses for you:
1. Initial reaction to your feature proposal
2. Suggested alternatives for porting your stored procedure

### 1) Initial reaction to your feature proposal

Based on what I understand, your primary proposition is to introduce the capability of managing exceptions, potentially by means of an additional hook named `when_exception` that would complement the [existing hooks](https://docs.getdbt.com/docs/build/hooks-operations#about-hooks).

Paraphrasing Drew's sentiment [here](https://github.com/dbt-labs/dbt-core/issues/1968#issuecomment-561244330) and making it my own:

We'd rather not add a new hook type to dbt unless it's truly impactful. After all, the best-case scenario would be if you could accomplish your transformations without a `when_exception` hook 😎 . I'm struggling to come up with scenarios that couldn't be addressed by an [existing materialization](https://docs.getdbt.com/docs/build/materializations#materializations) or a [custom materialization](https://docs.getdbt.com/guides/advanced/creating-new-materializations). However, I have faith that you or others will identify any scenarios that I may have overlooked 🙏

I definitely see how the logic of the stored procedure does have exception handling, but I _suspect_ it can be refactored to use dbt components that are currently available. Which brings us to...

### 2) Suggested alternatives for porting your stored procedure

#### TL;DR
Reading through the use case you described, it sounds like [`delete+insert` incremental strategy](https://medium.com/indiciumtech/understanding-dbt-incremental-strategies-part-2-2-add59889ea17#9fdc) could be one direction to consider. I don't pretend to understand your use-case well enough to know for certain, so you should check it out (and other [types of incremental materializations](https://medium.com/indiciumtech/understanding-dbt-incremental-strategies-part-1-2-22bd97c7eeb5)) for yourself!

#### General recommendations
In general, we advocate for declarative approach to defining a dataset rather than procedural flow of building a dataset. e.g. we encourage using built-in dbt materializations rather than a series of DDL and DML statements (even if there is an pre-existing stored procedure that works perfectly).

When using a built-in dbt materialization, I'd expect for a `when_exception` hook to be rarely necessary (but happy to learn about when it might be necessary!).

When porting an existing stored procedure, I'd consider these options in order, stopping at the first one that meets the requirements:
1. Is there an [existing built-in materialization](https://docs.getdbt.com/docs/build/materializations#materializations) that accomplishes the same goal?
    - See here for an in-depth look at different [strategies for incremental materializations](https://medium.com/indiciumtech/understanding-dbt-incremental-strategies-part-1-2-22bd97c7eeb5).
2. Can the logic be re-written into a series of models (like described [here](https://docs.getdbt.com/guides/migration/tools/migrating-from-stored-procedures/1-migrating-from-stored-procedures))?
3. Can the logic be accomplished via a custom `get_incremental_XYZ_sql` macro ([not yet documented](https://github.com/dbt-labs/docs.getdbt.com/issues/1761) 😬)?
4. Can the logic be accomplished in a general-purpose [custom materialization](https://docs.getdbt.com/guides/advanced/creating-new-materializations)?
5. If we got here, there are some extra tricky requirements... are they all truly necessary, or can the requirements be transformed so that options 1 or 2 can solve the problem?

I'd expect the first two to be the most common, and all the rest to be much more rare.

Could you give the suggested alternatives a look and see if any of them would work for you? If not, I'll be eager to learn what makes them not work.

**Comment by ddumasCT** [comment]
Hi Doug,

Thanks for getting back.  Is the delete+insert materialization wrapped inside a transaction?  If so, that should handle the functionality of DML, but still does not address actual exception handling.  As you know, exception handling has nothing to do with the type of materialization used.  However, do the custom alternatives you list allow for failure code that is triggered on DML failure?

I am surprised the lack of support for exception handling has not come up before.    As you know, It's built into all modern programming languages, and database developers rely on it.

I do feel that my idea of the when exception hook would add great value to the product.

Thanks,

Dave
________________________________

**Comment by jtcohen6** [comment]
@dbeatty10 said it well! I won't add too much more.

In terms of its philosophy, dbt is not a fully featured imperative-style programming language, nor do we think it should be. We try to avoid adding complexity unless it is solving for concrete use cases common in data/analytics engineering, and ideally in an opinionated way. While generic support for exception handling would be one way to solve for these use cases, we want to be thoughtful about addressing each in turn, and figuring out if it's truly the best way.

It sounds like the two use cases you called out are:
- Leaving the old unchanged table in place if the DML statement fails
- Logging information (metadata) about the failure

dbt already does both of these things, in a more standard & opinionated way:
- Materializations have "exception handling" in the sense of, rolling back open transactions (on data platforms that use transactions) if any statement fails. You declaratively define just the materialization & strategy you want to use, rather than rewriting imperative & boilerplate code (DDL/DML) for each model.
- dbt has a capable system for [structured logging](https://docs.getdbt.com/reference/events-logging), and writes [standard metadata artifacts](https://docs.getdbt.com/reference/artifacts/dbt-artifacts) at the end of every run. You can upload this metadata to the data warehouse using packages such as [`dbt_artifacts`](https://hub.getdbt.com/brooklyn-data/dbt_artifacts/latest/). This is preferable to, and more performant than, inserting single records into logging tables on analytical databases ([trust us, we tried that too!](https://github.com/dbt-labs/dbt-event-logging)).

**Comment by ddumasCT** [comment]
Hi Jeremy,

For the DML, we know that only the delete+insert  strategy will work, and not merge.  With delete+insert incremental strategy, does Dbt wrap both in a single transaction?  If not, and they are treated as 2 separate transactions that would be bad.  For example, the delete could succeed, and the subsequent insert could fail.  The table would then be left with deleted records = Not Good.

Can you elaborate on how Dbt already handles that.  I cannot find it anywhere in the doc.

Thanks,

Dave
________________________________

**Comment by jtcohen6** [comment]
> With delete+insert incremental strategy, does dbt wrap both in a single transaction?

On `dbt-snowflake`, yes!
- [Here's](https://github.com/dbt-labs/dbt-snowflake/blob/1dc2e7c60e20e925b95c909d4957f3909d2d2d03/dbt/include/snowflake/macros/adapters.sql#L275-L289) where we define the macro to wrap DML in explicit `begin;` + `commit;`
- [Here's](https://github.com/dbt-labs/dbt-snowflake/blob/f30ca40a596af646c749e6c82bf3a2209edc8502/dbt/include/snowflake/macros/materializations/merge.sql#L35-L38) where we use that macro to wrap `delete` + `insert` for that incremental strategy in particular

If you try it out, you should see dbt submitting `begin` + `commit` before and after the `delete` + `insert` statements.

**Comment by ddumasCT** [comment]
Jeremy,

Thanks for getting back  - reading the code it's not as clear to me.  I just want to make sure I understand. For the delete+insert incremental strategy, both DMLs will be in the same single transaction, correct?

Ex.

Begin;
      delete from where...;
      insert into ...;
Commit;

So, if the delete succeeds, and then the insert fails, the delete would be rolled back, correct?

Thanks,

Dave



________________________________

**Comment by ddumasCT** [comment]
Thanks Doug, I will try - I will force a failure with cast(my_varchar_column as date) function in the insert statement, and see if the delete is rolled back.
Dave
________________________________
